### PR TITLE
Scatterplot and Histogram: panning, pinching and RMB zooming works as expected.

### PR DESCRIPTION
**Pull request by BinarySoftware**
### Pull Request Description
2 finger panning on trackpad pans the plot, pinching and RMB zooms.

### Important Notes
GeoMap gestures will be fixed in a separate PR, asked a question about [custom events in MapBox on StackOverflow](https://stackoverflow.com/questions/66763635/mapbox-panning-map-with-2-fingers-on-a-trackpad)

### Checklist
Please include the following checklist in your PR:

- [ ] The `CHANGELOG.md` was updated with the changes introduced in this PR.
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [x] All code has automatic tests where possible.
- [x] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [x] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [ ] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 1]
```diff
@@ -254,11 +253,12 @@ class Histogram extends Visualization {
                 [0, 0],
                 [canvas.inner.width, canvas.inner.height],
             ])
-            .on(zoomClass, zoomed)
+            .on('zoom', zoomed)
+        // .on('mousewheel', zoomed)
```
Comment (MichaelMauderer) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: Commented code.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js [hunk 2]
```diff
@@ -173,16 +170,17 @@ class ScatterPlot extends Visualization {
...
      */
     addPanAndZoom(box_width, box_height, svg, margin, scaleAndAxis, scatter, points) {
-        let zoomClass = 'zoom'
-        let minScale = 0.5
-        let maxScale = 20
+        const minScale = 0.5
+        const maxScale = 20
+        const right_button = 2
+        const mid_button = 1
+        const scroll_wheel = 0
```
Comment (MichaelMauderer) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js]: Probably my fault, but the variable name convention differs between names. Should probably all be snakeCase. 
Maybe those constants should also be at the top of the file in the constants section.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js [hunk 3]
```diff
@@ -173,16 +170,17 @@ class ScatterPlot extends Visualization {
...
         const extent = [minScale, maxScale]
-        let zoom = d3
+        let start_pos
+        const zoom = d3
             .zoom()
             .filter(function () {
-                let right_button = 2
-                let mid_button = 1
-                let scroll_wheel = 0
+                console.log('filter', d3.event)
```
Comment (MichaelMauderer) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js]: Left over log statement.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js [hunk 4]
```diff
@@ -234,18 +267,28 @@ class ScatterPlot extends Visualization {
...
+        }
+
+        /**
+         * Return the zoom value computed from the initial right-mouse-button event to the current
+         * right-mouse event.
+         */
+        function rmb_zoom_value(event) {
+            const end = get_pos(event)
+            let delta_x = end.x - start_pos.x
+            let delta_y = end.y - start_pos.y
```
Comment (MichaelMauderer) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js]: Should all be `const`.

**Review by MichaelMauderer — CHANGES_REQUESTED** [review]
On Windows I am observing the following behaviour


| Interaction  | IDE  | ScatterPlot |
| ------------- | ------------- | ------------- |
| Two-finger up/down | Panning up/down| Zooming in/out  |
| Two-finger left/right | Panning left/right| Nothing* |
| RMB+move | Zooming in/out| Panning |

(*) Actually, I just noticed that this gesture might be captured by chrome: if I use two fingers and swipe left I toggle the "back" action of the browser. 

State is correctly preserved when switching interactions and using the fit-alll-button. That is wonderful!

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js (hunk 1 above)
Comment (BinarySoftware) [reply src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: I have no idea what it does here haha, i thought i've removed it ...

**Comment by MichaelMauderer** [comment]
@BinarySoftware Do you think there is a good way to abstract the code and avoid having the almost same code in both visualizations? We might want to add new visualizations in the future and having this as a drop-in feature would be really useful.

**Comment by BinarySoftware** [comment]
I've checked that on macOS & Windows machines and it works properly on both.

**Comment by wdanilo** [comment]
@BinarySoftware I would love to see it live. Please ping me after standup so we will talk about it.

**Comment by sylwiabr** [comment]
After visual review there were two problems for Scatterplot and Histogram plots:
1. for both plots zooming in was working wrong - it was not zooming in in proper place (i.e. zooming in for left top corner is zooming other part of the graph)
2. for histogram after zooming in the bars are looking bad - they don't show real results
3. the geomap still does not work

**Comment by BinarySoftware** [comment]
> After visual review there were two problems for Scatterplot and Histogram plots:
> 
> 1. for both plots zooming in was working wrong - it was not zooming in in proper place (i.e. zooming in for left top corner is zooming other part of the graph)
> 2. for histogram after zooming in the bars are looking bad - they don't show real results
> 3. the geomap still does not work

1 & 2 is fixed, 3. is left for another PR.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js [hunk 5]
```diff
@@ -65,6 +65,25 @@ function makeGenerator() {
...
+class MapController extends deck.MapController {
+    handleEvent(event) {
+        if (event.type === 'wheel') {
+            if (!event.srcEvent.ctrlKey) {
+                // this.controllerStateProps.latitude += event.srcEvent.deltaX
+                // this.controllerStateProps.longitude += event.srcEvent.deltaY
+                // this.updateViewport(this.controllerStateProps, {transitionDuration: 0}, {
+                //     isDragging: true,
+                //     isPanning: true
+                // });
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js]: commented code

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 6]
```diff
@@ -32,10 +32,10 @@ const BUTTON_HEIGHT = 25
...
    "focus" { "x" : 1.7, "y" : 2.1, "zoom" : 3.0 },
    "color" :  "rgb(1.0,0.0,0.0)" },
-   "bins" : 10,
-   "data" : [
-     "values" : [0.1, 0.2, 0.1, 0.15, 0.7],
-   ]
+ "bins" : 10,
+ "data" : [
+ "values" : [0.1, 0.2, 0.1, 0.15, 0.7],
+ ]
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: previous version had good foratting

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 7]
```diff
@@ -9,16 +9,16 @@ let shortcuts = {
...
+const labelStyle = 'font-family: DejaVuSansMonoBook; font-size: 10px;'
+const margin = 25
+const xAxisLabelWidth = 10
+const yAxisLabelWidth = 10
+const animationDuration = 1000
+const linearScale = 'linear'
+const lightPlotColor = '#00E890'
+const darkPlotColor = '#E0A63B'
+const defaultNumberOfBins = 10
+const buttonHeight = 25
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: why this change? We are using upper case for constants everywhere

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 8]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
                     case 'wheel':
-                        return d3.event.button === scroll_wheel
+                        return d3.event.button === scrollWheel
                     default:
                         return false
                 }
             })
+            .wheelDelta(function () {
+                const event = d3.event
+                // This is the delta used by default in wheelData and which is modified by us.
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: Could you elaborate more? What does it mean that it is used "by default", what is "wheelData", and how it is modified by us (what is the purpose?). Comment should answer those questions.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 9]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
+                const defaultWheelDelta =
+                    -event.deltaY * (event.deltaMode === 1 ? 0.05 : event.deltaMode ? 1 : 0.002)
+                return defaultWheelDelta * (event.ctrlKey ? 5 : 1)
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: magic numbers `5` and `1`

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 10]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
+                        return d3.event.button === scrollWheel
                     default:
                         return false
                 }
             })
+            .wheelDelta(function () {
+                const event = d3.event
+                // This is the delta used by default in wheelData and which is modified by us.
+                const defaultWheelDelta =
+                    -event.deltaY * (event.deltaMode === 1 ? 0.05 : event.deltaMode ? 1 : 0.002)
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: magic nubers, shuld be extracted to well named cosntants

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 11]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
         function zoomed() {
-            self.scale.zoom = d3.event.transform.k
-            let tmpScale = Object.assign({}, self.scale)
-            tmpScale.x = d3.event.transform.rescaleX(self.scale.x)
-            self.rescale(tmpScale, false)
+            // TODO:
+            // - When zooming, fix bins to show data properly
+            // - when panning, dont let user go less than 0 on Y scale, as it wont make sense.
+            if (d3.event.sourceEvent != null && d3.event.sourceEvent.buttons === rightButton) {
+                const zoomAmount = rmbZoomValue(d3.event.sourceEvent) / 5000.0
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: more magic numbers

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 12]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
+                const scale = Math.exp(zoomAmount)
+                const focus = startPos
+                const distanceScale = d3.zoomIdentity
+                    .translate(focus.x, focus.y)
+                    .scale(scale)
+                    .translate(-focus.x, -focus.y)
+                transformedScale.x = distanceScale.rescaleX(transformedScale.x)
+                // transformedScale.y = distanceScale.rescaleY(transformedScale.y)
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: more comented code

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 13]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
+                transformedScale.zoom = transformedScale.zoom * scale
+            } else if (d3.event.sourceEvent != null && d3.event.sourceEvent.type === 'wheel') {
+                if (d3.event.sourceEvent.ctrlKey) {
+                    const scale = Math.exp(-d3.event.sourceEvent.deltaY / 100.0)
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: even more magic nubers

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 14]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
+                    const focus = startPos
+                    const distanceScale = d3.zoomIdentity
+                        .translate(focus.x, focus.y)
+                        .scale(scale)
+                        .translate(-focus.x, -focus.y)
+                    transformedScale.x = distanceScale.rescaleX(transformedScale.x)
+                    // transformedScale.y = distanceScale.rescaleY(transformedScale.y)
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: even ore comented code

**Review by wdanilo — APPROVED** [review]
1. Super amazing its done!
2. The code contains a lot of coented code and agic numbers everywhere, this needs to be fixed.
3. I'd love to have a video chat to see it working today.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js (hunk 5 above)
Comment (MichaelMauderer) [reply src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js]: Also, this whole code block should probably be `if (combined_cond) { super.handleEvent(event) }` if the commented code is no longer relevant.

Comment (MichaelMauderer) [reply src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js]: If `GeoMap` is not fixed here, there probably should be no changes at all in this file.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js (hunk 7 above)
Comment (MichaelMauderer) [reply src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: Is this the `pretty` way?
I think we are actually missing linting for these JS files and there is an old PR of mine in limbo, which would have extended this.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js [hunk 15]
```diff
@@ -230,53 +230,124 @@ class Histogram extends Visualization {
...
          * Helper function called on pan/scroll.
          */
         function zoomed() {
-            self.scale.zoom = d3.event.transform.k
-            let tmpScale = Object.assign({}, self.scale)
-            tmpScale.x = d3.event.transform.rescaleX(self.scale.x)
-            self.rescale(tmpScale, false)
+            // TODO:
+            // - When zooming, fix bins to show data properly
+            // - when panning, dont let user go less than 0 on Y scale, as it wont make sense.
```
Comment (MichaelMauderer) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: Should these `TODOs` be resolved before merging?

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js [hunk 16]
```diff
@@ -172,83 +159,139 @@ class ScatterPlot extends Visualization {
...
-    addPanAndZoom(box_width, box_height, svg, margin, scaleAndAxis, scatter, points) {
-        let zoomClass = 'zoom'
-        let minScale = 0.5
-        let maxScale = 20
+    addPanAndZoom(boxWidth, boxHeight, svg, margin, scaleAndAxis, scatter, points) {
+        const minScale = 0.5
+        const maxScale = 20
+        const rightButton = 2
+        const midButton = 1
+        const scrollWheel = 0
```
Comment (MichaelMauderer) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:167]: Are these defined somewhere in D3 actually? If not maybe we should define them somewhere for use in all visualizations. Right now we just keep adding them in every file where we need them.

**Review by MichaelMauderer — CHANGES_REQUESTED** [review]

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js (hunk 5 above)
Comment (BinarySoftware) [reply src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js]: @MichaelMauderer I left it as super.handleEvent and added TODO, so that it will be easier for me or anyone to start working on next issue, knowing where to go.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js (hunk 16 above)
Comment (BinarySoftware) [reply src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:167]: No they're not, but i think that introducing another file in our configuration of loading strings from files to rust and then concatenating them is not a great idea.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js (hunk 8 above)
Comment (BinarySoftware) [reply src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js]: This is about speed of zooming in/out or panning, as described [here](https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent/deltaMode). I've fixed it of course by now its better.

File: src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js [hunk 17]
```diff
@@ -65,6 +65,24 @@ function makeGenerator() {
 
 const makeId = makeGenerator()
 
+/**
+ * Custom Map Controller, enabling us to redefine mouse gestures.
+ * TODO: Make 2-finger panning behave like in IDE, and RMB zooming.
```
Comment (MichaelMauderer) [inline src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/geoMap.js]: Can you link the TODO to an issue?

**Review by MichaelMauderer — COMMENTED** [review]
Code looks good. I still need to test this on a Windows machine, then I'll report back.

**Comment by wdanilo** [comment]
@MichaelMauderer , regarding oyu previous comment `Actually, I just noticed that this gesture might be captured by chrome: if I use two fingers and swipe left I toggle the "back" action of the browser.` - are you observing this behavior in GUI as well or only when visualization is focused? We are preventing this behavior in GUI and you should not see it ever. So if it shows up, this is a bug that should be fixed here.

**Comment by MichaelMauderer** [comment]
Looks good! 

> @MichaelMauderer , regarding oyu previous comment `Actually, I just noticed that this gesture might be captured by chrome: if I use two fingers and swipe left I toggle the "back" action of the browser.` - are you observing this behavior in GUI as well or only when visualization is focused? We are preventing this behavior in GUI and you should not see it ever. So if it shows up, this is a bug that should be fixed here.

I only observer this behaviour when running the IDE in Chrome and doing two finger swipes to the left while the visualization is active.  It does not happen if the visualization is not active. In that case, the IDE pans as expected. So, either we need to fix this here or open a new issue.

**Comment by wdanilo** [comment]
@MichaelMauderer This issue is about no proper panning and zooming in visualizations, so yes, 2 finger gesture not working on Windows is part of the issue. However, what do you mean by "Chrome"? Does it work in Electron but fails in Chrome?

If this fails only in Chrome but works in Electron, we can indeed extract it to a separate issue. It should be still covered here, but due to lack of tie we can do it this way.

**Comment by BinarySoftware** [comment]
> @MichaelMauderer This issue is about no proper panning and zooming in visualizations, so yes, 2 finger gesture not working on Windows is part of the issue. However, what do you mean by "Chrome"? Does it work in Electron but fails in Chrome?
> 
> If this fails only in Chrome but works in Electron, we can indeed extract it to a separate issue. It should be still covered here, but due to lack of tie we can do it this way.

Opened an issue for that: https://github.com/enso-org/ide/issues/1378

**Review by MichaelMauderer — APPROVED** [review]
